File: spell-checker/kotlin-implementation/app/src/main/kotlin/io/github/srolfe26/vozeexercise/SpellChecker.kt
package io.github.srolfe26.vozeexercise

import java.io.File

class SpellChecker(private val dictionaryPath: String) {
    private val dictionary: Set<String> = loadDictionary()

    private fun loadDictionary(): Set<String> {
        return File(dictionaryPath).useLines { lines ->
            lines.map { it.trim().lowercase() }.toSet()
        }
    }

    fun isWordValid(word: String): Boolean {
        if (word.isEmpty()) return false
        val lowercaseWord = word.lowercase()

        return when {
            isSingleLetterWord(lowercaseWord) -> true
            lowercaseWord.contains("'") -> {
                val firstPart = lowercaseWord.split("'")[0]
                isSingleLetterWord(firstPart) || dictionary.contains(firstPart)
            }
            else -> dictionary.contains(lowercaseWord)
        }
    }

    private fun isSingleLetterWord(word: String): Boolean =
        word.length == 1 && (word == "a" || word == "i")

    fun getSuggestions(word: String, maxSuggestions: Int = 5, maxDistance: Int = 2): List<String> {
        if (word.isEmpty() || dictionary.contains(word.lowercase())) return emptyList()

        return dictionary
            .filter { Math.abs(it.length - word.length) <= maxDistance }
            .map { it to damerauLevenshteinDistance(word.lowercase(), it) }
            .filter { (_, distance) -> distance <= maxDistance }
            .sortedBy { (_, distance) -> distance }
            .take(maxSuggestions)
            .map { (suggestion, _) -> suggestion }
    }

    fun isPotentialProperNoun(word: String): Boolean {
        if (word.isEmpty() || !word[0].isUpperCase()) return false
        val lowercaseWord = word.lowercase()
        if (dictionary.contains(lowercaseWord)) return true
        return getSuggestions(word, 1, 1).isEmpty()
    }

    fun checkText(filePath: String): List<MisspelledWord> {
        val content = File(filePath).readText()
        val lines = content.split("\n")

        return lines.flatMapIndexed { lineNum, line ->
            val words = Regex("\\b\\w+(?:'\\w+)?\\b").findAll(line)
            words.flatMap { matchResult ->
                val word = matchResult.value
                val columnNum = matchResult.range.first
                if (!isWordValid(word) && !isPotentialProperNoun(word)) {
                    listOf(
                        MisspelledWord(
                            word,
                            lineNum + 1,
                            columnNum + 1,
                            line.substring(
                                (columnNum - 20).coerceAtLeast(0),
                                (columnNum + word.length + 20).coerceAtMost(line.length)
                            ),
                            getSuggestions(word)
                        )
                    )
                } else {
                    emptyList()
                }
            }
        }
    }

    private fun damerauLevenshteinDistance(a: String, b: String): Int {
        val m = a.length
        val n = b.length
        val d = Array(m + 1) { IntArray(n + 1) }

        for (i in 0..m) d[i][0] = i
        for (j in 0..n) d[0][j] = j

        for (i in 1..m) {
            for (j in 1..n) {
                val cost = if (a[i - 1] == b[j - 1]) 0 else 1
                d[i][j] = minOf(
                    d[i - 1][j] + 1,  // deletion
                    d[i][j - 1] + 1,  // insertion
                    d[i - 1][j - 1] + cost  // substitution
                )
                if (i > 1 && j > 1 && a[i - 1] == b[j - 2] && a[i - 2] == b[j - 1]) {
                    d[i][j] = minOf(d[i][j], d[i - 2][j - 2] + cost)  // transposition
                }
            }
        }

        return d[m][n]
    }
}

data class MisspelledWord(
    val word: String,
    val line: Int,
    val column: Int,
    val context: String,
    val suggestions: List<String>
)

fun main(args: Array<String>) {
    if (args.size != 2) {
        println("Usage: kotlin SpellChecker.kt <dictionary_file> <text_file>")
        return
    }

    val (dictionaryPath, textPath) = args
    val spellChecker = SpellChecker(dictionaryPath)

    val misspelledWords = spellChecker.checkText(textPath)

    if (misspelledWords.isNotEmpty()) {
        println("Misspelled words:")
        misspelledWords.forEach { (word, line, column, context, suggestions) ->
            println("\n\"$word\" at line $line, column $column")
            println("Context: ...${context}...")
            println("Suggestions: ${suggestions.joinToString(", ")}")
        }
    } else {
        println("No misspelled words found.")
    }
}

File: spell-checker/node-js-implementation/spell-checker.js
const fs = require("fs").promises;

const loadDictionary = async (filePath) => {
  const content = await fs.readFile(filePath, "utf-8");
  return new Set(content.split("\n").map((word) => word.trim().toLowerCase()));
};

const isSingleLetterWord = (word) =>
  word.length === 1 && (word === "a" || word === "i");

const isWordValid = (dictionary) => (word) => {
  if (typeof word !== "string" || word.length === 0) {
    return false;
  }

  const lowercaseWord = word.toLowerCase();

  if (isSingleLetterWord(lowercaseWord)) {
    return true;
  }

  if (lowercaseWord.includes("'")) {
    const firstPart = lowercaseWord.split("'")[0];
    return isSingleLetterWord(firstPart) || dictionary.has(firstPart);
  }
  return dictionary.has(lowercaseWord);
};

const getSuggestions =
  (dictionary) =>
  (word, maxSuggestions = 5, maxDistance = 2) => {
    if (typeof word !== "string" || word.length === 0) {
      return [];
    }
    const lowercaseWord = word.toLowerCase();

    // Return an empty array if the word is already in the dictionary
    if (dictionary.has(lowercaseWord)) {
      return [];
    }

    return Array.from(dictionary)
      .filter(
        (dictWord) =>
          Math.abs(dictWord.length - lowercaseWord.length) <= maxDistance
      )
      .map((dictWord) => ({
        word: dictWord,
        distance: damerauLevenshteinDistance(lowercaseWord, dictWord),
      }))
      .filter(({ distance }) => distance <= maxDistance)
      .sort((a, b) => a.distance - b.distance)
      .slice(0, maxSuggestions)
      .map((suggestion) => suggestion.word);
  };

const isPotentialProperNoun = (dictionary, getSuggestionsFunc) => (word) => {
  if (typeof word !== "string" || word.length === 0) {
    return false;
  }
  if (word[0] !== word[0].toUpperCase()) {
    return false;
  }
  const lowercaseWord = word.toLowerCase();
  if (dictionary.has(lowercaseWord)) {
    return true; // Consider it a proper noun if its lowercase version is in the dictionary
  }
  const suggestions = getSuggestionsFunc(word, 1, 1);
  return suggestions.length === 0;
};

const damerauLevenshteinDistance = (a, b) => {
  const m = a.length;
  const n = b.length;
  const d = Array(m + 1)
    .fill()
    .map(() => Array(n + 1).fill(0));

  for (let i = 0; i <= m; i++) {
    d[i][0] = i;
  }
  for (let j = 0; j <= n; j++) {
    d[0][j] = j;
  }

  for (let i = 1; i <= m; i++) {
    for (let j = 1; j <= n; j++) {
      const cost = a[i - 1] === b[j - 1] ? 0 : 1;
      d[i][j] = Math.min(
        d[i - 1][j] + 1, // deletion
        d[i][j - 1] + 1, // insertion
        d[i - 1][j - 1] + cost // substitution
      );
      if (i > 1 && j > 1 && a[i - 1] === b[j - 2] && a[i - 2] === b[j - 1]) {
        d[i][j] = Math.min(d[i][j], d[i - 2][j - 2] + cost); // transposition
      }
    }
  }

  return d[m][n];
};

const checkText =
  (isWordValidFunc, isPotentialProperNounFunc, getSuggestionsFunc) =>
  async (filePath) => {
    const content = await fs.readFile(filePath, "utf-8");
    const lines = content.split("\n");

    return lines.flatMap((line, lineNum) => {
      const words = line.match(/\b\w+(?:'\w+)?\b/g) || [];
      let columnNum = 0;

      return words.flatMap((word) => {
        columnNum = line.indexOf(word, columnNum);
        const isValid =
          isWordValidFunc(word) || isPotentialProperNounFunc(word);
        const result = !isValid
          ? [
              {
                word,
                line: lineNum + 1,
                column: columnNum + 1,
                context: line.substring(
                  Math.max(0, columnNum - 20),
                  Math.min(line.length, columnNum + word.length + 20)
                ),
                suggestions: getSuggestionsFunc(word),
              },
            ]
          : [];
        columnNum += word.length;
        return result;
      });
    });
  };

const main = async () => {
  if (process.argv.length !== 4) {
    console.error("Usage: node spell_checker.js <dictionary_file> <text_file>");
    process.exit(1);
  }

  const [, , dictionaryPath, textPath] = process.argv;

  try {
    const dictionary = await loadDictionary(dictionaryPath);
    const isWordValidWithDict = isWordValid(dictionary);
    const getSuggestionsWithDict = getSuggestions(dictionary);
    const isPotentialProperNounWithDict = isPotentialProperNoun(
      dictionary,
      getSuggestionsWithDict
    );
    const checkTextWithFunctions = checkText(
      isWordValidWithDict,
      isPotentialProperNounWithDict,
      getSuggestionsWithDict
    );

    const misspelledWords = await checkTextWithFunctions(textPath);

    if (misspelledWords.length > 0) {
      console.log("Misspelled words:");
      misspelledWords.forEach(
        ({ word, line, column, context, suggestions }) => {
          console.log(`\n"${word}" at line ${line}, column ${column}`);
          console.log(`Context: ...${context}...`);
          console.log(`Suggestions: ${suggestions.join(", ") || "None"}`);
        }
      );
    } else {
      console.log("No misspelled words found.");
    }
  } catch (error) {
    console.error("An error occurred:", error.message);
    process.exit(1);
  }
};

if (require.main === module) {
  main();
} else {
  module.exports = {
    loadDictionary,
    isWordValid,
    isPotentialProperNoun,
    getSuggestions,
    damerauLevenshteinDistance,
    checkText,
  };
}
